core/resharding.go

```diff
@@ -127,17 +133,48 @@ func GetShardingStateFromBlockChain(bc *BlockChain, epoch uint64) *ShardingState
 
 // CalculateNewShardState get sharding state from previous epoch and calculate sharding state for new epoch
 // TODO: currently, we just mock everything
-func CalculateNewShardState(bc *BlockChain, epoch uint64) types.ShardState {
+func CalculateNewShardState(bc *BlockChain, epoch uint64, stakeInfo *map[common.Address]*structs.StakeInfo) types.ShardState {
 	if epoch == FirstEpoch {
-		return fakeGetInitShardState(6, 10)
+		return getInitShardState(3, 10)
 	}
 	ss := GetShardingStateFromBlockChain(bc, epoch-1)
-	newNodeList := fakeNewNodeList(int64(ss.rnd))
+	newNodeList := ss.UpdateShardingState(stakeInfo)
 	percent := ss.calculateKickoutRate(newNodeList)
-	ss.UpdateShardState(newNodeList, percent)
+	ss.Reshard(newNodeList, percent)
 	return ss.shardState
 }
 
+// UpdateShardingState remove the unstaked nodes and returns the newly staked node Ids.
+func (ss *ShardingState) UpdateShardingState(stakeInfo *map[common.Address]*structs.StakeInfo) []types.NodeID {
+	oldAddresses := make(map[common.Address]bool)
+	for _, shard := range ss.shardState {
+		newNodeList := shard.NodeList[:0]
+		for _, nodeID := range shard.NodeList {
+			addr := common.Address{}
+			addrBytes, err := hex.DecodeString(string(nodeID))
+			if err != nil {
+				utils.GetLogInstance().Error("Failed to decode address hex")
+			}
+			addr.SetBytes(addrBytes)
+			oldAddresses[addr] = true
+			_, ok := (*stakeInfo)[addr]
+			if !ok {
```

Comment:
it should be changed to "if ok" if it means the addr is still in stake?

Reply:
Good catch. Fixed.